feat(nav): add overflow styling to horizontal nav examples

christiemolloy opened the pull request:
fix #1519 

- Is it possible for the box shadow to change to white in the dark theme?

Visuals:
<img width="748" alt="Screen Shot 2019-03-22 at 10 00 03 AM" src="https://user-images.githubusercontent.com/20118816/54827874-5150ef80-4c89-11e9-9834-150a0ad011be.png">

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit 63468c2d01478b09faa515acdce5a483b936d9af

https://deploy-preview-1615--pf-next.netlify.com

mcoker commented:
Does this component's scroll buttons need the current styles, like used in tabs?

christiemolloy commented:
@mcoker I don't think so as its just meant to be an addition to overflow scrolling but I can check

mcoker commented:
Since the UI and basic interaction is the same, seems like they should match in that regard. Also, since this pattern is effectively the same between the components, I'm curious if this should be it's own component that nav and tabs use. And maybe it has a modifier to create both the UI used for the primary and secondary tabs examples.

christiemolloy commented:
In terms of a new component would you be talking about the overflow button as a component or the entire horizontal list overflow. I don't know how I feel about that because one is a nav and one is a secondary tab list, and the lists have really different styles. and then we have the horizontal nav version for the masthead that seems like it should be an example of the nav component rather than using a component. I could also see the styles for secondary tabs maybe changing too @mcoker

mcoker commented:
Looks like it would be an overflow-scroll component that you put a list inside of, and it could apply to tabs, secondary tabs, horizontal navigation, or if you had another list somewhere you wanted to treat similarly. Just in general, it's the same structure, and if I came across this pattern in the nav and tabs, I would expect it to behave the same or at least very similarly. The visual differences could be handled in the choice of icon or modifiers applied. It just looks like there is a lot of copy and paste between the two currently. What I would want to avoid is having 2 unique implementations because there are slight design changes between them. And question the design differences anyways, and ask if, since they're very similar patterns, they should present themselves more similarly, too.

I don't think we need to worry about it for this PR, but something to think about in the future.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -550,3 +585,34 @@
+  padding-right: var(--pf-c-nav__scroll-button--PaddingRight);
+  padding-left: var(--pf-c-nav__scroll-button--PaddingLeft);
+  background-color: var(--pf-c-nav__scroll-button--BackgroundColor);
+  border: initial;
```
Is there a reason you use `initial` over `0`?

christiemolloy replied:
This was used in the tab example so I thought it was purposeful. I can make a note to update both.

christiemolloy commented:
@mcoker the reason I moved `overflow-x: auto` to the horizontal/tertiary list elements is because on the nav component itself, it was causing the buttons to scroll with the list. But placing it on the list fixes this.
Also something strange is happening in Firefox: the button isn't positioning absolute to its parent, its extending a certain amount below (hard to see), I don't think its because we're hiding the scrollbar with width: none, because this hasn't been a problem before.

<img width="621" alt="Screen Shot 2019-03-26 at 10 57 51 AM" src="https://user-images.githubusercontent.com/20118816/55007903-942b0400-4fb6-11e9-8e90-f0e31377be65.png">

mcoker commented:
Just want to verify that the right-side scroll button should extend to the edge of the viewport like this?

<img width="1714" alt="Screen Shot 2019-03-26 at 1 57 17 PM" src="https://user-images.githubusercontent.com/35148959/55025588-32c35f00-4fcf-11e9-8cfc-b5c021b87344.png">

Also when I place the horizontal navigation in the page header nav, I can't scroll and the scroll button isn't visible.

<img width="1714" alt="Screen Shot 2019-03-26 at 2 02 53 PM" src="https://user-images.githubusercontent.com/35148959/55025918-ddd41880-4fcf-11e9-8cb9-2782eed20c1a.png">

mcoker commented:
And this is how tertiary nav looks with both scroll buttons visible.

<img width="1544" alt="Screen Shot 2019-03-26 at 2 05 11 PM" src="https://user-images.githubusercontent.com/35148959/55026080-399ea180-4fd0-11e9-9057-50fd3f65da12.png">

christiemolloy commented:
`--pf-c-page__main-nav--PaddingRight: 0` is assigned to 0 for some reason

christiemolloy commented:
@mcoker updated this, visuals have approved of the box shadows.
Two things:
- The box shadow for light and dark are hard coded. I didn't make them global vars because I think its a one off (i.e. we dont have a var for a white box-shadow or one that is the same color as dark mode), but that means the light/dark theming doesn't apply here and I had to add the dark var to the component...which I dont think we want.
- Also I added a media query to the tertiary-list-item so that the first child has a padding-left of 32px on desktop, and then a padding-left of 16px on mobile. This looks really good, but I don't think we want to handle the padding in the component, and it would definitely be better for `.pf-c-page-main-nav` to handle the padding....so maybe we do explore negative positioning on the button?

I also added some styles to the page to show how the nav would work.. Ill remove those, but definitely check out the nav on the page component.

christiemolloy commented:
@mcoker fyi if you click on "the second option for nav" I added a different variation for this option that adds negative margin to the buttons

mcoker commented:
Re: box shadows, that's fine with me. Are we sure we don't want to introduce dark mode box shadows?

I like option 2. by...

> also what if the .pf-c-page__main-nav handled the pf-c-nav__scroll-button , since its a one off case.

Do you mean that element handles the negative positioning on the buttons? If so, I think that's ideal. Considering this should be a modular component, it should follow the rules of the component and the parent should handle the positioning for one-off stuff like this.

Though just tossing it out there, what if the buttons didn't go so far into the negative space, and the icon aligned with the content?

<img width="520" alt="Screen Shot 2019-03-27 at 4 50 28 PM" src="https://user-images.githubusercontent.com/35148959/55114930-42b36f80-50b1-11e9-8361-6361b445f1fc.png">

side note - do you know why this and the tabs button use `width` and don't let the padding determine the width?

christiemolloy commented:
@mcoker I successfully added the horizontal nav to the masthead in the page demo & updated the box shadow color as per @mceledonia  recommendation

mcoker commented on src/patternfly/components/Nav/docs/code.md:
```diff
@@ -24,6 +24,27 @@ The navigation system relies on several different sub-components:
 
 | Class | Applied To | Outcome |
 | -- | -- | -- |
+<<<<<<< HEAD
```
Looks like you have some merge remnants left in here...

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -164,20 +164,46 @@
 
-  // Hide scrollbars
-  &::-webkit-scrollbar {
+  &.pf-m-dark {
```
I wonder if instead of `pf-m-dark`, you should just style the button `box-shadow` in the horizontal nav? And can the color and background colors not inherit from `@include pf-t-dark;` in `.pf-c-nav__horizontal-list`?

mcoker commented on src/patternfly/components/Page/page.scss:
```diff
@@ -204,6 +215,10 @@
   overflow: hidden;
   background-color: var(--pf-c-page__header-nav--BackgroundColor);
 
+  > * {
```
is there a reason you're using `*` instead of `.pf-c-nav`?

christiemolloy replied:
I was thinking for flexibility but I don't think that's necessary here so good point!

mcoker commented on src/patternfly/demos/Page/page-template-title.hbs:
```diff
@@ -1,3 +1,38 @@
+{{#> page-main-nav page-main-section--modifier="pf-m-light"}}
```
Are you sure you want to add tertiary nav to every demo? Also the expandable nav already had tertiary nav, so now you have 2 in that demo - https://deploy-preview-1615--pf-next.netlify.com/demos/Page/examples-full/?component=Page%20Component%20Expandable%20Nav%20Example

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -164,20 +164,33 @@
+  --pf-c-nav__scroll-button--m-dark--nth-of-type-2--BoxShadow: -20px 0 10px -4px rgba(21, 21, 21, .7); // This is a special one-off shadow for nav
+
   // Base styling
-  overflow-x: auto;
```
Sorry if I asked you this already, but why move this from nav to the horizontal nav components? Curious if this is necessary?

mcoker commented:
@christiemolloy this looks great! Just one question about moving the `overflow` property from the main nav component into the horizontal nav variations.

Also, I see you are not using pf-m/t-dark, or including `%pf-t-dark` anywhere. Is the only remaining issue with the examples? If so, you can update the example component for those 2 examples to also include the `pf-t-dark` class.


```
      <Example
        heading="Horizontal Nav (only in masthead)"
        handlebars={navHorizontalListExampleRaw}
        className="is-dark-preview pf-t-dark"
      >
        {navHorizontalListExample}
      </Example>
      <Example
        heading="Horizontal Nav Overflow (only in masthead)"
        handlebars={navHorizontalListOverflowExampleRaw}
        className="is-dark-preview pf-t-dark"
      >
```

mcoker commented:
Ah, I see what you're saying about needing a dark class on the nav if the buttons are outside of the list. In this case, we style another nav element based on the page element it's applied in, and I think that's fine here, too. The buttons have to be outside of the list though, otherwise you get a11y errors. Plus if we make this a component in the future, the buttons will need to be outside of the list (and probably the entire nav element). 

In this PR - https://github.com/christiemolloy/patternfly-next/pull/13 I refactored that and fixed a couple of a11y issues, as well as used `aria-label="Global"` on the main nav components in the demos, to match what's on react and what was on master before this PR. I'm assuming that was lost in a merge conflict? I'm not sure about whether to use `aria-label="Global"` or `aria-label="Local"` on the tertiary nav in the demos though. The tertiary nav in the current page demo on master uses "Global" but the tertiary nav in the Nav component examples uses "Local". Maybe @jgiardino can guide us here?

christiemolloy commented:
@mcoker thanks for making those changes! This is happening in the example because the page__header-nav has 'width: auto` on it rather than width: 100%, but it works in the demo, so not sure if it even matters:

<img width="560" alt="Screen Shot 2019-04-04 at 10 43 19 AM" src="https://user-images.githubusercontent.com/20118816/55564848-cd562900-56c6-11e9-8d93-16b20d6b8631.png">

mcoker commented:
@christiemolloy I think that's fine. This is one of those cases where the example doesn't look right because it's outside of the context where it is designed to be used. Though looks like you could add `.pf-c-page__header` as a wrapper to those examples and that fixes it.

jgiardino commented:
@mcoker I'm not seeing what you're describing. 

We're using what's called **Horizontal Nav**  not **Tertiary Nav** in our Page component. And on the the Nav component page, it does have `aria-label="Global"`, which is correct.

![image](https://user-images.githubusercontent.com/21063328/55567628-0f359e00-56cc-11e9-8c43-e5bb57b2e338.png)

I did notice that before this PR, the page component has two aria-label attributes though. One of them was changed to "Global" in this PR, so now at least they're both correct, but we only need one. 😉

jgiardino commented on src/patternfly/demos/DataList/data-list-page-nav.hbs:
```diff
@@ -1,4 +1,4 @@
-{{#> nav nav--attribute=(concat 'id="' page--id '-primary-nav" aria-label="Grouped Nav Example"')}}
+{{#> nav nav--attribute=(concat 'id="' page--id '-primary-nav" aria-label="Global"')}}
```
`aria-label` appears to already be defined on the `nav` handlebar. It's showing up twice in the Page and Datalist demos.

mcoker replied:
Nice find! Looks like we should just remove any reference to `aria-label` from all of the nav components in the examples, except for the tertiary nav used in the page expandable demo, data list demo, and data table demos. And in those cases, define `aria-label` as an attribute of the hbs partial, not in the `nav--attribute` attribute value, since this is the logic that generates `aria-label` in `nav.hbs`

```
<nav class="pf-c-nav{{#if nav--modifier}} {{nav--modifier}}{{/if}}"
  {{#if aria-label}}
    aria-label="{{aria-label}}"
  {{else}}
    aria-label="Global"
  {{/if}}
  {{#if nav--attribute}}
    {{{nav--attribute}}}
  {{/if}}>
  {{> @partial-block}}
</nav>
```

mcoker replied:
Or alternatively, we could remove the `aria-label` logic from `nav.hbs`, and include it in `nav--attribute` instead. That would probably be more consistent with our other partials.

mcoker commented:
@jgiardino we use tertiary nav on the [expandable nav page demo](https://deploy-preview-1615--pf-next.netlify.com/demos/Page/examples-full/?component=Page%20Component%20Expandable%20Nav%20Example). 

<img width="1390" alt="Screen Shot 2019-04-04 at 10 28 38 AM" src="https://user-images.githubusercontent.com/35148959/55568045-89622480-56c4-11e9-8196-3e9e50c8a03a.png">

It's also in all of the [data list demos](https://pf4.patternfly.org/demos/DataList/examples/) and [data table demos](https://pf4.patternfly.org/demos/DataTable/examples/)

jgiardino commented:
> we use tertiary nav on the [expandable nav page demo](https://deploy-preview-1615--pf-next.netlify.com/demos/Page/examples-full/?component=Page%20Component%20Expandable%20Nav%20Example).

Thanks for the example, @mcoker! That example uses the correct aria-labels. The primary nav is referred to as "global" navigation. Then the tertiary nav region is "local" navigation relative to the "Resource Usage" page that's selected in the global navigation.

mcoker commented on src/patternfly/demos/DataList/data-list-page-nav.hbs:
```diff
@@ -1,4 +1,4 @@
-{{#> nav nav--attribute=(concat 'id="' page--id '-primary-nav" aria-label="Grouped Nav Example"')}}
+{{#> nav nav--attribute=(concat 'id="' page--id '-primary-nav" aria-label="Local"')}}
```
```suggestion
{{#> nav nav--attribute=(concat 'id="' page--id '-primary-nav" aria-label="Global"')}}
```

mcoker commented:
Left one change suggestion.

Also. Can you just remove the nav component from this page component example, and replace the component with `pf-c-nav` text like the other examples?

https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/components/Page/examples/page-component-nav-vertical-example.hbs#L17-L19

I also wonder if we should add `aria-label` to the nav component examples? It's on some of them but not all of them. We can create a follow up issue to do that, too.

patternfly-build commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly)
- [GitHub release](https://github.com/patternfly/patternfly-next/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
